docs/taskinputvalidation.md

```diff
@@ -0,0 +1,135 @@
+2. *Non goal/Goal to get holistic view* - Provide task input enforcement and also ability to advise possible input values
+    - Eg: If a task input called `destination` is `abc` then the possible list of values for a task input called `target` should be
+    *one of* `[x,y,z]`. This imples that the value of the input `target` should be enforced to be one of them *and* also the UI should advise those inputs as possible values to the user.
+3. *Non goal/Unclear goal* - Authors should be able to run custom javascript
+    - Task authors can point to a javascript file and render the whole task editor in iframe?
+    - Task authors can point to custom javascript to control just task input's visibility and validation?
+
+-------
+
+Supported boolean functions
+-------
+
+> `value` is alias to task input value
+
+- `isUrl(value: string)` - true if value is of valid URL, allows ftp urls as well
+- `isIpV4Address(value: string)` - true if value is a valid IPV4 address
+- `isEmail(value: string)` - true if value is of email format
+- `isEmpty(value: string)` - true if value is empty
+- `isInRange(value: string, min: number, max: number)` - true if value is <= max and >= min
+- `isSha1(value: string)` - true if value is a valid sha1 hash
+- `isWinUsername(value: string)` - true if value is of samAccountName or userPrincipalName format
+- `isLowerCase(value: string)` - true if value is all lower case
+- `isUpperCase(value: string)` - true if value is all upper case
+- `isMinlength(value: string, min: number)` - true if length is greater than or equal to number
+- `isMaxlength(value: string, max: number)` - true if length is less than or equal to number
+- `isPath(value: string)`- true if value has valid windows NTFS path characters
+- `isMatch(value: string, regEx: string, regExOptions: string)` - true if value matches the regex
+    - Since validation could be performed on C# and javascript, this is how we can achieve that:
+        - We would be using `ECMAScript` for C# [regex options](https://msdn.microsoft.com/en-us/library/system.text.regularexpressions.regexoptions(v=vs.110).aspx)
+        - When we use `ECMAScript`, we can only specify `IgnoreCase` and/or `Multiline` options to C# regex
+        - With such limitations, javascript equivalent flags we support would be `g` (for global match, this gets translated to getting a [single match](https://msdn.microsoft.com/en-us/library/system.text.regularexpressions.regex.match(v=vs.110).aspx) or [all matches](https://msdn.microsoft.com/en-us/library/system.text.regularexpressions.regex.matches.aspx) in C#), `i` (for C# `IgnoreCase`), `m` (for C# `Multiline`)
+        - Default C# behavior would be to use single match
+        - So, we would be supporting these options:
+            - IgnoreCase (i)
+            - Multiline (m)
+        - Valid options - "im", "m", "mi", "i"
```

Comment:
I would prefer `IgnoreCase,Multiline` rather than abbreviations. Abbreviations are bad when someone else needs to maintain the code. Abbreviations are good for things that people need to interactively type and want a shorthand for (e.g. common command line instructions), for things that others may need to maintain, prefer readability.